osd, mon, extblkdev: add support for thin provisioned NVMe devices

ifed01 opened the pull request:
This allows to use NVMe devices with compression on-board in Ceph and see the utilization stats properly via Ceph means

Apart from high level changes in monitor/osd code this patch relies on EBD plugin infra which was introduced for thin-provisioning devices. New plugin 'thin_nvme' has been introduced.

<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins test classic perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-classic/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test crimson perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-crimson/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test signed` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pr-commits/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-commits/config/definitions/ceph-pr-commits.yml)
- `jenkins test make check` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests/config/definitions/ceph-pull-requests.yml)
- `jenkins test make check arm64` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests-arm64/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests-arm64/config/definitions/ceph-pull-requests-arm64.yml)
- `jenkins test submodules` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-pr-submodules/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-submodules/config/definitions/ceph-pr-commits.yml)
- `jenkins test dashboard` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-pull-requests/config/definitions/ceph-dashboard-pull-requests.yml)
- `jenkins test dashboard cephadm` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-cephadm-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-cephadm-e2e/config/definitions/ceph-dashboard-cephadm-e2e.yml)
- `jenkins test api` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-api/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-api/config/definitions/ceph-pr-api.yml)
- `jenkins test docs` [ReadTheDocs](https://readthedocs.org/projects/ceph/) | [Github Workflow Definition](https://github.com/ceph/ceph/blob/main/.readthedocs.yml)
- `jenkins test ceph-volume all` [Jenkins Jobs](https://jenkins.ceph.com/view/ceph-volume%20PR/) | [Jenkins Jobs Definition](https://github.com/ceph/ceph-build/blob/main/ceph-volume-cephadm-prs/config/definitions/ceph-volume-pr.yml)
- `jenkins test windows` [Jenkins Job](https://jenkins.ceph.com/job/ceph-windows-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-windows-pull-requests/config/definitions/ceph-windows-pull-requests.yml)
- `jenkins test rook e2e` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-orchestrator-rook-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-rook-e2e/config/definitions/ceph-orchestrator-rook-e2e.yml)
</details>

ifed01 committed:
blk,extblkdev,bluestore: move statfs update logic into ebd plugin

Signed-off-by: Igor Fedotov <igor.fedotov@croit.io>

ifed01 committed:
extblkdev: add thin_nvme plugin

Signed-off-by: Igor Fedotov <igor.fedotov@croit.io>

ifed01 committed:
osd,mon,extblkdev: Use store_statfs_t::*_raw members for tracking physical device stats

Signed-off-by: Igor Fedotov <igor.fedotov@croit.io>

ifed01 committed:
osd: simplify PG::pg_stat_adjust

Signed-off-by: Igor Fedotov <igor.fedotov@croit.io>

ifed01 committed:
osd: modify full ratio calculation to use *raw stats

Signed-off-by: Igor Fedotov <igor.fedotov@croit.io>

ifed01 committed:
mon/pgmap: dump avail_raw along with used_raw.

Signed-off-by: Igor Fedotov <igor.fedotov@croit.io>

ifed01 committed:
mon/pgmap: fix reweight-by-utilization to use *_raw settings

Signed-off-by: Igor Fedotov <igor.fedotov@croit.io>

ifed01 committed:
mon, osd: adjust using regular vs. raw statfs parameters

Signed-off-by: Igor Fedotov <igor.fedotov@croit.io>

ifed01 committed:
.gitmodules: add libnvme submodule

Signed-off-by: Igor Fedotov <igor.fedotov@croit.io>

ifed01 commented:
kenkinst test dashboard

ifed01 commented:
jenkins test windows

ifed01 committed:
blk/kernel, extblkdev: add nvme device autodiscovery for thin_nvme vdo plugin.

Phisical NVMe device might sit behind symbolic name, LVM, etc.
This patch allows the plugin to discover proper underlying NVMe namespace if any.

Signed-off-by: Igor Fedotov <igor.fedotov@croit.io>

ifed01 commented:
jenkins test make check

ifed01 commented:
jenkins test make check arm64

ifed01 commented:
jenkins test windows

ifed01 commented:
jenkins test dashboard

ifed01 commented:
ceph test dashboard

aclamk commented on src/osd/osd_types.h:
```diff
@@ -2439,12 +2439,15 @@ bool operator==(const pg_stat_t& l, const pg_stat_t& r);
  */
 struct store_statfs_t
 {
```
#0: 
How about us documenting roles of `store_statfs_t` field?
My proposal that I think reflects reality:

`total` = Disk capacity, usually as declared by manufacturer and reported by driver; this value influences crush calculation.
`available` = How much more information can this device hold; at deploy it should be ~equal to `total`.
When compression is disabled `available` decreases in the rate of `data_stored` rising. When compression is enabled and data is compressible `available` decreases slower than `data_stored`. This value is used by balancer.
`internally_reserved` = I suppose this is the amount of capacity used for ObjectStore management of the device; it is 0 for both BlueStore and SeaStore.
`allocated` = ObjectStore specific value that reflects how much of disk capacity is used; for BlueStore and SeaStore it reflects disk blocks allocated.
`data_stored` = Size of data in objects. It might be less than allocated when objects are small, or it might be more than allocated if snapshots and clones are present.

`data_compressed_original` = BlueStore only. Size of data in objects that underwent compression.
`data_compressed_allocated` = BlueStore only. Size of disk allocated for compressed objects. Whenever this is updated, `allocated` is updated too.
`data_compressed` = BlueStore only. Size (in bytes) obtained after compression of data tracked by `data_compressed_original`.

`omap_allocated` = BlueStore only. Amount of capacity used specifically for omaps.
`internal_metadata` = BlueStore only. Size of all BlueFS that is not `omap_allocated`.


For the PR notes:
#1:
I am against renaming `allocated`->`data_allocated`. First, a lot of changes little gain; second, if anything, it should be `disk_allocated`.
#2:
`total_raw` and `avail_raw` should have better explanations
#3
Do we even need additional fields in `store_statfs_t`?
Based on https://docs.redhat.com/en/documentation/red_hat_enterprise_linux/7/html/storage_administration_guide/vdo-ig-integrating-sysfs-entries
and
https://github.com/torvalds/linux/blob/master/include/linux/nvme.h#L435-L437
we should just have a special calculation of `available`, 
and all specifics map via `collect_metadata()`.

aclamk commented on src/os/bluestore/BlueStore.cc:
```diff
@@ -12107,21 +12113,6 @@ void BlueStore::_get_statfs_overall(struct store_statfs_t *buf)
-    // underlying physical device.
-    bfree = std::min(bfree, ebd_state.get_physical_avail());
-
-    buf->allocated = ebd_state.get_physical_total() - ebd_state.get_physical_avail();;
```
Interesting, this `buf->allocated` was overwritten in `BlueStore::statfs()`.

ifed01 replied:
true, but apparently nobody cared...

aclamk commented on src/extblkdev/thin_nvme/ExtBlkDevThinNVMe.cc:
```diff
@@ -0,0 +1,124 @@
+  } else {
+    uint64_t lba_size = 1ULL << ns_data.lbaf[ns_data.flbas & 0x0F].ds;
+
+    uint64_t nsze = ns_data.nsze * lba_size;
```
It would be nice to comment `nsze` `ncap` `nuse`, or at least link to nvme documentation that explain it.

ifed01 committed:
extblkdev: enable effective caps settings if requested.

Set effective caps if requested by extblkdev plugin as well.
Originally permitted ones were updated only which is insufficient to
ioctl nvme devices.

Signed-off-by: Igor Fedotov <igor.fedotov@croit.io>

ifed01 committed:
osd: improving comments on store_statfs_t fields

Signed-off-by: Igor Fedotov <igor.fedotov@croit.io>

afreen23 reviewed (approved):
dashboard changes

MartinVerges commented:
Please merge
